### ISSUE TITLE
FR: Apply current defaults to this document

### ISSUE THREAD

**Issue by offset-torque** [koreader/koreader#7708]
**PROBLEM DEFINITON**
(This might not be a common problem for average reader but it affects users like me)
I am an academic reader so at any moment I am reading many books (~10) and articles (~5) in parallel.

> **SCENARIO:** Occasionally I come across a better font than I am currently using so I want to change my documents to it. Changing font generally includes changing text size, boldness, contrast, word spacing and line spacing too. So now I have to remember and change these 6 properties (with ~15-20 clicks, I counted) for each of the 15 open documents (~200-350 clicks in total). And if I come across a better font later or get bored from the current one: **Repeat this again :(((**

**OFFERED SOLUTION**
A menu button with this function: **"Apply current defaults to this document"**
So I can set my changes as defaults in one document and update others with just a click.

**ALTERNATIVE SOLUTION**
Moon Reader solves this problem with THEMES. A theme includes all visual options. So you can save your settings as a theme and switch to it any time. 

With this solution users can update a DEFAULT THEME in koreader and it will apply to all documents automatically.

This solution adds an extra usability advantage. For example we can add a NIGHT THEME which has: 
- bigger and bolder font
- more line spacing 
- warmer backlight
- inverted text color

Of course this might be harder to implement than the "Apply current defaults to this document" solution I offered above (which probably includes just updating the current document's metadata.lua).

**Additional context**
Koreader (and its DEVS) is awesome. It is like the BLENDER of e-book reading. Many thanks to everyone working on it !!!

**Comment by NiLuJe** [comment]
There's a Profiles plugin that aims to resolve at least some of these pain points.

(It's tied into the Gesture manager, so, anything you can do there can be used to build a Profile).

Since you've talked about fonts, while I know that it handles font sizes (because it came up recently ;p), I don't quite recall if we can set a font *family* this way...

**Comment by NiLuJe** [comment]
Note that I took the nerd-fi approach to this issue: I mangle my usual fonts to have compatible metrics ;p.

**Comment by hius07** [comment]
> text size, boldness, contrast, word spacing and line spacing

Additionally resets page margins.

**Comment by hius07** [comment]
For wider discussion. https://github.com/koreader/koreader/pull/8412#discussion_r742306840

I would separate per documents settings into 4 groups:

(1) view settings (font, page)
We often play with these settings to achieve the best view of the document, it would be nice to have an option to reset these settings all together to default, **being within a book**. Hence this issue.
Are there any pdf-document settings to be reset here?

(2) highlights and bookmarks
There is `Bulk remove` in the Bookmark list to reset.

(3) style tweaks
Maybe it is useful to have a menu item to reset all tweaks? Don't know, I change them rarely.

(4) miscellaneous
Includes various settings from different menus and dialogs, hard to combine. They are important but not being changed very often or massively, so there is no need to reset them all together.

**Comment by offset-torque** [comment]
**Thank you so much @hius07 for spending effort on another issue that I opened.**

### 1.
Biggest offender here is font change. Since so many details change from font to font (glyph width, ascenders/descenders, baseline, thickness, serifs/ligatures), it generally needs a chain of adjustments after changing the font.

For an (extreme) example going from **Lato** to **Noto Serif** needs:
- Increased font size
- Increased font weight/contrast
- Increased line spacing
- Maybe increased word spacing

to maintain same level of  readability.

**Hinting, kerning, render mode, zoom (dpi)** options has a **"set to best and forget"** nature as we recommend in the user guide. So they can be left out of resetting I guess.

**PDF:** PDF files can be so diverse (margins/columns layout, text style/placement) I think it is impossible to set defaults that would work for every use case. I have to tweak nearly each PDF individually to make it readable on this small screen, despite mainly reading scientific journals (which should be the widely used standard two-column layout).

Some of them play nice with reflow, others turn into a mess. Margins can change even between pages. And most importantly we can't change the font in PDF's so chain of changes mentioned above are not needed. We don't have EPUB flexibility here. Actually if REFLOW is off we don't have anything to customize. So not much here for defaults and reset (in my experience).

### 2.
**Highlights and bookmarks:** I never needed to reset them. We have also **purge .sdr** option in FM. That does the job I guess. **Purge .sdr** also resets the document to defaults. So all use cases are covered when combined with **Bulk remove** you mentioned (as far as I understand).

### 3.
**Style tweaks:** I only use them to globally remove publisher formatting and never touch them. I know some users use them heavily and they can be very document specific. So I can't comment about this use case.

### 4.
I agree with you.

---
**SUMMARY:** I think your implementation will cover majority of use cases that needs this type of reset (definitely covers my proposal). Placement and naming of buttons in the menus are logical and clear also.

**Comment by hius07** [comment]
I agree with https://github.com/koreader/koreader/pull/8412#discussion_r742306840 that `Reset font and page settings` must be available from the current opened book.
(What is the best place for a menu item? The first navi tab, below Toc and Bookmarks? Or the last hamburger tab, near Book info and Book status? Or in the typeset tab, near Style and Font?)
(Or (extremely) in the first tab of the bottom menu?)

If so, should it be also available from the file browser and history (as proposed in https://github.com/koreader/koreader/pull/8412)?
Or resetting the current book only is enough?

**Comment by poire-z** [comment]
I'd say deep down in `Gear > Document > `. It's not something essential that should be first level.
And for me, it does not need to be anywhere else than in the current book Reader menu.

(Reset settings in the History/File browser menus are mostly used - by me at least - to just unmark a book as opened, and you get the info it is opened or not with the dogear. You wouldn't get any info about previously opened books needing page and font settings reset, or already have it done, in there.)

**Comment by offset-torque** [comment]
> I agree with #8412 (comment) that Reset font and page settings must be available from the current opened book.

I agree with this too.
General use case will be probably like this:
- User changes some defaults in current book
- User opens another book to read and sees that it has **"old style"**
- User resets book

So user is already **in** the book with **"old style"** to notice that it needs resetting.

> If so, should it be also available from the file browser and history (as proposed in #8412)?

Only use case I can think for this is if a user wants to batch reset some books. But **"resetting after opening the book for reading"** is same effort as that. So an **"in book menu item"** is enough I think. Less places to change, less things to break.

> I'd say deep down in Gear > Document > . It's not something essential that should be first level.

I don't agree with this. This feature might not be frequently used but when it is used it will be **repeatedly** used. User will do this action for every (reflowable) document he already opened to apply his new style. I think it should be on a first level menu to avoid unnecessary nesting clicks. And of course discoverability goes out the window when it is tucked deep down in some sub-menu.

**For placement:** Since this option resets the "style" of document, users would (logically) expect to find this feature in the menu where most style options are. Maybe just below **"Style tweaks"**. Here users can notice they have this option available when they want to change their font which is **exactly when this feature is needed**.

**Comment by Frenzie** [comment]
I concur that below "typography rules" would make sense.

**Comment by hius07** [comment]
![11](https://user-images.githubusercontent.com/62179190/140310401-c86da7fb-b3c9-4c98-b172-42d58bca81cc.png)

![12](https://user-images.githubusercontent.com/62179190/140310411-9169b21e-a973-4e0b-b668-bd6da1b47948.png)

**Comment by offset-torque** [comment]
I was going to propose first position like this but I thought it might get protested on the basis of "non-frequent feature shouldn't be the first item". So I think this is perfect 💯

**Comment by poire-z** [comment]
> User will do this action for every (reflowable) document he already opened to apply his new style. I think it should be on a first level menu to avoid unnecessary nesting clicks. 

I think he/you'll do it when you open an old book and notice it has an old style. I don't think you'll ever go on a campaign of opening all your old books to apply this new feature (because the re-rendering time of each book will fed you up - and the additional tap will feel like nothing compared to that).

> exactly when this feature is needed

This feature is needed by you :) who is font-changed focused.
My issue is that we don't really agree with what should be reset above:
- @hius07 adds everything page/font related
- @offset-torque feels _Hinting, kerning, render mode, zoom (dpi) options has a "set to best and forget" nature as we recommend in the user guide. So they can be left out of resetting I guess._
- I don't need any of these, but I feel that if we allow to reset _some_ page/font related settings, we should allow resetting more/all kinds of per-document settings - even if we don't do it right now, we should work on this @offset-torque 's wish so it can be extended to some other use cases - which is why having it in a first level typeset menu feels a bit wrong to me - and Gear>Document> would feel better and more generic.

But may be we won't need anything more and if you all feel it's right there in that menu under Typography Rules... well, ok.
But then I still think it should be granular (with radio buttons allowing to select page/margins, font-related, allowing the various sets that @hius07 and @offset-torque wished as default) - or full layout stuff (including hyphenation, typography, hyphenation) - so one does not need to wonder what did or did not get reset.

(Somehow, I don't like "Reset view settings to default" at the top. I'm so used to that menu that it feels alien and in my face :( But it would feel odd at the bottom I guess - at top, it's less reachable and express it targets everything else in that menu.)

**Comment by offset-torque** [comment]
@poire-z

> @offset-torque feels Hinting, kerning, render mode, zoom (dpi) options has a "set to best and forget" nature as we recommend in the user guide. So they can be left out of resetting I guess.

I said this based on your (Frenzie, NiLuJe, you) answers in the user guide thread. You can re-read what everyone said there if you forgot.

---

Typography menu is not crowded. And if you still want to throw something under the most unused sub-menu, let me introduce you to the first item here **STYLE** !!! 

I personally asked this question in the user guide thread:
**"Is there any scenario that a user may need to choose a Style/CSS other than AUTO ?"**

**NiLuJe's answer:**
> Probably not.

**Your answer:**
> epub.css is picked for EPUBs. But html5.css might be prefered on some books.

So there is no need to change and you can't even tell me as a developer that in which cases a user has to change STYLE. Then how does users know when to change it ? Of course they don't know. So they don't touch this. They just leave it at AUTO like 99% of users. **So why an item that is never touched is the first item there ?**

> I'm so used to that menu that it feels alien and in my face :( 

@poire-z feels weird so no one should touch that menu. Your feelings are more important than the improvement of the software I guess.

**Comment by NiLuJe** [comment]
I see tact is still not your strong suit ;).

**Comment by hius07** [comment]
![13](https://user-images.githubusercontent.com/62179190/140315529-fdc3aadb-9ea5-4eae-acc6-74431499409c.png)

**Comment by hius07** [comment]
With reflow disabled there's almost nothing to reset for pdf, maybe `kopt_contrast`?
Or just hide the menu item?

**Comment by poire-z** [comment]
With PDF, as I'm not really familiar with all the options, I had in the past felt a bit lost with the Fit/Zoom/Crop/Scroll mode (possibly before they were moved to the bottom menu) - so I think having the possibility to reset all these can be helpful widh PDF/DJVU.

**Comment by Frenzie** [comment]
@offset-torque 
> but I thought it might get protested on the basis of "non-frequent feature shouldn't be the first item".

It very much will be protested on that basis. It doesn't or at the very least shouldn't make sense on top.

It could be worth investigating why you feel it should receive such prominent billing as there might be a better way to solve what underlies the desire.

> So why an item that is never touched is the first item there ?

Historically, because it once made more sense. But today, because it's logical to place it together with style tweaks and reversing those two would look less pretty and also unexpected. But an argument against the location of Style is *not* an argument *for* putting Reset on top. ;-) It's just an argument for swapping font stuff and style stuff around. Reset logically goes below those, regardless in which order you put them. ^_^

@hius07 
> With reflow disabled there's almost nothing to reset for pdf, maybe kopt_contrast?
Or just hide the menu item?

There are extensive cropping, contrast, etc. options that you might wish to reset.

**Comment by poire-z** [comment]
@offset-torque : we had our fights in the past, and I guess it's hard for both of us to read the other without any prejudice... but as we are on the topic of resetting, can we reset us ?

When I mention you, I am not (always:) attacking you and your wishes - quite the contrary if you read me well: I was just stating that we (hius07, you, me) had different initial expectations on what should be reset, and I did not say any of them was better or worse than the others.
And I just expressed my feeling (it was not (yet :) a veto) about the menu item - like you did - so that others can take it into consideration, like they did yours.

**Comment by offset-torque** [comment]
@Frenzie 
> It could be worth investigating why you feel it should receive such prominent billing as there might be a better way to solve what underlies the desire.

**My reasoning was like this:** Other items in the menu change one aspect of text like **PAGE OVERLAP** or **HIGHLIGHT STYLE**. This feature is like a sledgehammer that changes 8 properties of text at once irreversibly. There is no undo. So it seems kind of important (dangerous).

But of course I am not insisting on any order as long as it is on a first level menu. I just voted +1 for hius07's decision to point out that we thought similarly.

---

@poire-z 
> we had our fights in the past, and I guess it's hard for both of us to read the other without any prejudice... but as we are on the topic of resetting, can we reset us ?

I too agree with this sentiment and definitely not enjoying this energy draining type of conversation. Thank you for explaining your perspective. I am sure you are a good person but we started badly and now the problem is (as you pointed out) we are reading each other with prejudice. I will make effort to take your replies in good faith. I hope you will do the same.

**Comment by hius07** [comment]
For pdf

![14](https://user-images.githubusercontent.com/62179190/140327219-9078348d-6896-4aed-a5db-a4b45eaab99e.png)

**Comment by Frenzie** [comment]
Does reflow mode implicitly mean all the reflow mode settings as well? They're not accessible when reflow is disabled, but they are set up in a certain way. Also, watermark doesn't depend on reflow.

I think I kind of skipped some discussion earlier, but to me reset logically means probably everything other than bookmarks/highlights.

**Comment by hius07** [comment]
> Does reflow mode implicitly mean all the reflow mode settings as well?

Default reflow state is off, so other settings enabled in reflow mode are not touched.



> but to me reset logically means probably everything other than bookmarks/highlights.

We (I hope) have agreed that not all settings to be reset (font/page only which are often played with).

**Comment by Frenzie** [comment]
As the person actually putting in the work your opinion holds above average weight, but I'd work with a blacklist (i.e., preciously few properties not deleted), not a whitelist (i.e., it's already horrifically long and it feels like a hundred things are still missing).

**Comment by hius07** [comment]
For reflowable docs the list is comprehensive (based on this issue).

I am not very experienced in reading pdf, are there any other settings to be reset?
One opinion:


> **PDF:** PDF files can be so diverse (margins/columns layout, text style/placement) I think it is impossible to set defaults that would work for every use case. I have to tweak nearly each PDF individually to make it readable on this small screen, despite mainly reading scientific journals (which should be the widely used standard two-column layout).
> 
> Some of them play nice with reflow, others turn into a mess. Margins can change even between pages. And most importantly we can't change the font in PDF's so chain of changes mentioned above are not needed. We don't have EPUB flexibility here. Actually if REFLOW is off we don't have anything to customize. So not much here for defaults and reset (in my experience).

**Comment by offset-torque** [comment]
Let me clarify a bit @Frenzie (at least my issue opening reason). This is a screenshot of MoonReader. Probably the most widely used e-book reader app in the world (play store stops counting after 10 million so I can't be sure).

This current issue is actually a substitute of MoonReader's theme support. A theme includes visual items seen in the image below. Users can change a theme so all books automatically changes (which hius07 is trying to implement here). As you can see, hius07's current implementation covers all items in the screenshot (and some more). 

My proposal is following 80/20 rule and covering majority of users. If a user needs more items included in the RESET list like **"Render Mode"** really badly, they can open an issue about it just like I did. So instead of trying to cover all edge cases we can go **"Good enough"** and make additions later if it proves necessary.

![moonreader_theme](https://user-images.githubusercontent.com/84420582/140332327-bd7a0a36-abe3-4b5c-a5ae-c129e1c99817.jpg)

**Comment by Frenzie** [comment]
> My proposal is following 80/20 rule and covering majority of users. If a user needs more items included in the RESET list like "Render Mode" really badly, they can open an issue about it just like I did. So instead of trying to cover all edge cases we can go "Good enough" and make additions later if it proves necessary.

This argument is overlooking the fact that "good enough" takes significantly more effort. It's therefore worse in every conceivable way.

**Comment by poire-z** [comment]
> > to me reset logically means probably everything other than bookmarks/highlights.

> We (I hope) have agreed that not all settings to be reset (font/page only which are often played with).

Not sure I did :) I was with @Frenzie 's idea in https://github.com/koreader/koreader/pull/8412#discussion_r742276638 .

"_font/page only which are often played with_" is now a bit restrictive if the menu item ends up being at the top of the typeset menu, which contains Typography, Style tweaks, Hyphenation: not having it reset these feels then even more arbitrary (cf the various different opinions I mentionned earlier).

Rather than discussing which settings to reset, with the menu item there, it's probably easier to go with a blacklist of what to not reset:
- bookmarks  / highlights
- current position (xpointer/page)
- may be others ?

**Comment by Frenzie** [comment]
> Rather than discussing which settings to reset, with the menu item there, it's probably easier to go with a blacklist of what to not reset:

Exactly. QFT.

**Comment by poire-z** [comment]
> may be others ?

A quick read at some metadata.epub.lua, and it feels there is not many to not reset:
`bookmarks_sorted`, `highlights_imported`
`bookmarks_version` (didn't remember we already had some kind of bookmark migration :),
`cre_dom_version` (because it needs some migration when reset)

**Comment by hius07** [comment]
Okay, I'll make a "keep" list.

**Comment by hius07** [comment]
Reset all except kept.

https://github.com/koreader/koreader/blob/ee077ced7155821fbca61f32dbf3a2eaf1758cdc/frontend/apps/filemanager/filemanagerutil.lua#L42-L51


![15](https://user-images.githubusercontent.com/62179190/140500573-e10e8006-a79a-42ff-bed2-60075448ab32.png)
